src/dc_imap.rs

```diff
@@ -428,6 +438,52 @@ impl Imap {
                     )
                 };
 
+                return 0;
+            }
+        };
+
+        match login_res {
+            Ok((mut session, stream)) => {
+                // TODO: error handling
+                let caps = session.capabilities().unwrap();
+                let can_idle = caps.has("IDLE");
+                let has_xlist = caps.has("XLIST");
+
+                let caps_list = caps.iter().fold(String::new(), |mut s, c| {
+                    s += " ";
+                    s += c;
+                    s
+                });
+                let caps_list_c = std::ffi::CString::new(caps_list).unwrap();
+
+                info!(context, 0, "IMAP-capabilities:%s", caps_list_c.as_ptr());
+
+                let mut config = self.config.write().unwrap();
+                config.can_idle = can_idle;
+                config.has_xlist = has_xlist;
+                config.addr = Some(addr.into());
+                config.imap_server = Some(imap_server.into());
+                config.imap_port = Some(imap_port.into());
+                config.imap_user = Some(imap_user.into());
+                config.imap_pw = Some(imap_pw.into());
+                config.server_flags = Some(server_flags);
+
+                *self.session.lock().unwrap() = (Some(session), Some(stream));
+
+                1
+            }
+            Err((err, _)) => {
+                eprintln!("failed to login: {:?}", err);
+
+                unsafe {
+                    dc_log_event_seq(
+                        context,
+                        Event::ERROR_NETWORK,
+                        &mut 0 as *mut i32,
+                        b"Cannot login\x00" as *const u8 as *const libc::c_char,
```

Comment:
what's this double-as meaning?

Reply:
I think it's just a remnant from the conversion and not technically needed.

Reply:
This is to convert a rust static string into a c style pointer string, you can see it in many places in the code from the conversion.